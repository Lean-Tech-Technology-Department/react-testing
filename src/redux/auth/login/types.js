export const LOGIN_TYPES = {
  LOGIN_USER: '@LOGIN/LOGIN_USER',
  LOGIN_USER_SUCCESS: '@LOGIN/LOGIN_USER_USER_SUCCESS',
  LOGIN_USER_FAILURE: '@LOGIN/LOGIN_USER_FAILURE',
  LOGOUT: '@LOGIN/LOGOUT'
}
