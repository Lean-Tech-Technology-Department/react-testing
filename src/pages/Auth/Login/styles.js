import styled from 'styled-components'

const LoginPage = styled.div`
  height: 100vh;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  .form {
    width: 100%;
    max-width: 350px;
    padding: 10px 40px 40px 40px;
    box-shadow: 1px 1px 5px gray;
    border-radius: 5px;
  }

  .input {
    margin: 15px 0px;
  }
`

export { LoginPage }