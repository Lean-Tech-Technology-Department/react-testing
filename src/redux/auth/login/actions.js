export const LOGIN_TYPES = {
  LOGIN_USER: '@LOGIN/LOGIN_USER',
  SUCCESS_LOGIN_USER: '@LOGIN/SUCCESS_LOGIN_USER',
  FAILURE_LOGIN_USER: '@LOGIN/FAILURE_LOGIN_USER'
}
