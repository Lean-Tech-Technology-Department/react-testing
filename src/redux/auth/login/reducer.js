import { LOGIN_TYPES } from './types'
export const initialState = {
  user: {},
  loading: false,
  error: ''
}

export default (state = initialState, { type, payload }) => {
  switch (type) {
    case LOGIN_TYPES.LOGIN_USER: {
      return { ...state, user: {}, loading: true }
    }
    case LOGIN_TYPES.LOGIN_USER_SUCCESS: {
      return { ...state, user: payload, loading: false }
    }
    case LOGIN_TYPES.LOGIN_USER_FAILURE: {
      return { ...state, error: payload, loading: false }
    }
    default: {
      return state
    }
  }
}
