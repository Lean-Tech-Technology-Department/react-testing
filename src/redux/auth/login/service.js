import { Fetch } from '../../../utils/fetch'
import { routes } from './routes'

const login = (payload) => {
  return Fetch('post', routes.login, payload)
}

export { login }
