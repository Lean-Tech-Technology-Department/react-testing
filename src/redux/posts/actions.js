import { POSTS_TYPES } from './types'

export const getPosts = (onSuccess, onError) => ({
  type: POSTS_TYPES.GET_POSTS
})

export const getPostsSuccess = (payload) => {
  return {
    type: POSTS_TYPES.GET_POSTS_SUCCESS,
    payload
  }
}

export const getPostsFailure = (payload) => {
  return {
    type: POSTS_TYPES.GET_POSTS_FAILURE,
    payload
  }
}
export const createPosts = (payload, onSuccess, onError) => ({
  type: POSTS_TYPES.CREATE_POST, payload, onSuccess, onError
})

export const createPostSuccess = (payload) => {
  return {
    type: POSTS_TYPES.CREATE_POST_SUCCESS,
    payload
  }
}

export const createPostFailure = (payload) => {
  return {
    type: POSTS_TYPES.CREATE_POST_FAILURE,
    payload
  }
}

export const updatePost = (payload, onSuccess, onError) => ({
  type: POSTS_TYPES.UPDATE_POST, payload, onSuccess, onError
})

export const updatePostSuccess = (payload) => {
  return {
    type: POSTS_TYPES.UPDATE_POST_SUCCESS,
    payload
  }
}

export const updatePostFailure = (payload) => {
  return {
    type: POSTS_TYPES.UPDATE_POST_FAILURE,
    payload
  }
}

export const deletePost = (payload, onSuccess, onError) => ({
  type: POSTS_TYPES.DELETE_POST, payload, onSuccess, onError
})

export const deletePostSuccess = (payload) => {
  return {
    type: POSTS_TYPES.DELETE_POST_SUCCESS,
    payload
  }
}

export const deletePostFailure = (payload) => {
  return {
    type: POSTS_TYPES.DELETE_POST_FAILURE,
    payload
  }
}
