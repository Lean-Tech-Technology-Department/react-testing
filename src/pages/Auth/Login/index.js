import React, { useState } from 'react'
import { TextField, Button, Typography, CircularProgress } from '@material-ui/core'
import { LoginPage } from './styles'
import { useDispatch, useSelector } from 'react-redux'
import { login } from '../../../redux/auth/login/actions'

const Login = () => {
  const dispatch = useDispatch()
  const { loading } = useSelector(state => state.loginReducer)
  const [error, setError] = useState('')
  const [credentials, setCredentials] = useState({
    username: '',
    password: ''
  })

  const handleChange = (name) => (event) =>
    setCredentials((prevState) => ({
      ...prevState,
      [name]: event.target.value
    }))

  const handleLogin = () => {
    if (credentials.username && credentials.password) {
      setError('')
      dispatch(login(credentials))
    } else {
      setError('The username and the password are required')
    }
  }

  return (
    <LoginPage>
      <form className="form" autoComplete="off">
        <h1>Welcome!</h1>
        <div className="input">
          <TextField
            required
            id="username"
            value={credentials.username}
            onChange={handleChange('username')}
            label="Username"
            placeholder="Username"
            variant="outlined"
            inputProps={{
              'data-testid': 'loginUsernameInput'
            }}
            style={{ width: '100%' }}
          />
        </div>
        <div className="input">
          <TextField
            required
            id="password"
            value={credentials.password}
            onChange={handleChange('password')}
            label="Password"
            placeholder="Password"
            variant="outlined"
            inputProps={{
              'data-testid': 'loginPasswordInput'
            }}
            type="password"
            style={{ width: '100%' }}
          />
        </div>
        {Boolean(error) && <Typography role="error" variant="subtitle1" gutterBottom color="error">{error}</Typography>}
        <Button variant="contained" color="primary" onClick={handleLogin} style={{ width: '100%' }}>
          {loading ? <CircularProgress color="secondary" size={20} /> : 'Login'}
        </Button>
      </form>
    </LoginPage>
  )
}

export default Login
