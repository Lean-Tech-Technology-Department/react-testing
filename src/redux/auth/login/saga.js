import { takeEvery, call, put, all } from 'redux-saga/effects'
import { LOGIN_TYPES } from './types'
import * as loginActions from './actions'
import { login } from './service'

export function* onLogin({ payload }) {
  try {
    const response = yield call(login, payload)
    yield put(loginActions.loginSuccess(response))
  } catch ({ message }) {
    yield put(loginActions.loginFailure(message))
  }
}

function * onHandleLogin() {
  yield takeEvery(LOGIN_TYPES.LOGIN_USER, onLogin)
}

export function * loginSaga() {
  yield all([
    call(onHandleLogin)
  ])
}
