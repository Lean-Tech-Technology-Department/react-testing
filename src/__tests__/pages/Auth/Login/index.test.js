import React from 'react'
import { fireEvent, waitFor } from '@testing-library/react'
import * as loginActions from '../../../../redux/auth/login/actions'
import { renderWithProviders, TestProviders } from '../../../../utils/tests'
import Login from '../../../../pages/Auth/Login'
import { credentialsMock } from '../../../../utils/tests/mocks'

const mockDispatch = jest.fn()
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockDispatch
}))

describe('Login component', () => {
  it('renders snapshot correctly', () => {
    const { asFragment } = renderWithProviders(<Login />)
    expect(asFragment(<Login />, { wrapper: TestProviders })).toMatchSnapshot()
  })

  it('handles the inputs validation', async() => {
    const { getByText, getByRole } = renderWithProviders(<Login />)
    const loginButton = getByText('Login')
    fireEvent.click(loginButton)
    await waitFor(() => {
      expect(getByRole('error')).toHaveTextContent('The username and the password are required')
    })
  })

  it('handles the login action', async() => {
    const { getByTestId, getByText } = renderWithProviders(<Login />)
    const usernameInput = getByTestId('loginUsernameInput')
    const passwordInput = getByTestId('loginPasswordInput')
    const loginButton = getByText('Login')
    fireEvent.change(usernameInput, { target: { value: credentialsMock.username } })
    fireEvent.change(passwordInput, { target: { value: credentialsMock.password } })
    fireEvent.click(loginButton)
    await waitFor(() => {
      expect(mockDispatch).toHaveBeenCalledWith(loginActions.login(credentialsMock))
    })
  })
})
