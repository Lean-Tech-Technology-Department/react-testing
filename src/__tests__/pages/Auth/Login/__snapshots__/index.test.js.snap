// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component renders snapshot correctly 1`] = `
<DocumentFragment>
  <div>
    <div
      class="sc-bdvvaa gWiPQm"
    >
      <form
        autocomplete="off"
        class="form"
      >
        <h1>
          Welcome!
        </h1>
        <div
          class="input"
        >
          <div
            class="MuiFormControl-root MuiTextField-root"
            style="width: 100%;"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
              data-shrink="false"
              for="username"
              id="username-label"
            >
              Username
              <span
                aria-hidden="true"
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input"
                data-testid="loginUsernameInput"
                id="username"
                placeholder="Username"
                required=""
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-3"
                >
                  <span>
                    Username *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="input"
        >
          <div
            class="MuiFormControl-root MuiTextField-root"
            style="width: 100%;"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
              data-shrink="false"
              for="password"
              id="password-label"
            >
              Password
              <span
                aria-hidden="true"
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input"
                data-testid="loginPasswordInput"
                id="password"
                placeholder="Password"
                required=""
                type="password"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-3"
                >
                  <span>
                    Password *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
          style="width: 100%;"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Login
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
